---
title: "BLASTx of unmapped contigs to TE proteins"
output: rmarkdown::github_document
editor_options: 
  markdown: 
    wrap: sentence
    author: Riccardo
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(svglite))
theme_set(theme_bw())
```


```
blastx -query full.output.fa -subject ../transposons-proteins/transposons-clean.fasta -outfmt 6 > blast/TE-proteins.tsv
```


```{r}
blastx <- read_tsv("/Volumes/Storage/mining/GenomeDelta2.0/output/blast/TE-proteins.tsv", col_names = c("contig", "protein", "similarity", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore"))
```

```{r}
(best_match <- blastx %>%
   group_by(contig) %>%
   filter(bitscore==max(bitscore)) %>%                        # Keep best match with a protein in the database
   ungroup() %>%
   filter(bitscore>100, similarity > 30) %>%                  # Filter low bitscores and low similarity
   distinct() %>%
   mutate(contig_sep = contig) %>%
   separate(contig_sep, into = c("node","contig_n","l","contig_length","cov","contig_coverage"), sep="_") %>%
   select(-l, -node, -cov) %>%
   type_convert() %>%
   filter(contig_coverage > 5, contig_length < 15000))        # Filter low coverage and huge contigs
```

```{r}
(best_match %>% select(contig) %>% distinct() %>% write_tsv("/Volumes/Storage/mining/GenomeDelta2.0/output/TE-contigs.txt", col_names = FALSE))
```


